File: medium/2044. Count Number of Maximum Bitwise-OR Subsets/2044. Count Number of Maximum Bitwise-OR Subsets.go
package main

import "fmt"

func countMaxOrSubsets(nums []int) int {
	maxBitwiseOr := 0
	res := 0

	for _, num := range nums {
		maxBitwiseOr |= num
	}

	fmt.Println(maxBitwiseOr)

	var backtrack func(index, orValue int)

	backtrack = func(index, orValue int) {
		if orValue == maxBitwiseOr {
			res += 1
		}

		for i := index; i < len(nums); i++ {
			backtrack(index+1, orValue|nums[i])
		}
	}

	backtrack(0, 0)

	return res
}

File: medium/2044. Count Number of Maximum Bitwise-OR Subsets/2044. Count Number of Maximum Bitwise-OR Subsets.py
from typing import List


class Solution:
    def __init__(self) -> None:
        self.res = 0
        self.max = 0
        self.nums = []
        self.N = 0

    def countMaxOrSubsets(self, nums: List[int]) -> int:
        for num in nums:
            self.max |= num

        self.nums = nums
        self.N = len(nums)
        self.backtrack(0, 0)
        return self.res

    def backtrack(self, index: int, or_value: int):
        if or_value == self.max:
            self.res += 1

        for i in range(index, self.N):
            self.backtrack(i + 1, or_value | self.nums[i])


print(Solution().countMaxOrSubsets(nums=[3, 1]))
print(Solution().countMaxOrSubsets(nums=[2, 2, 2]))
print(Solution().countMaxOrSubsets(nums=[3, 2, 1, 5]))
